### PR TITLE
Add Key value store based on a typescript port of the keyv package

### PR DESCRIPTION
**Pull request by nklomp**
This is a Key Value store with a typescript port of the Keyv package (https://github.com/jaredwray/keyv)., ESM compatible and should work in browser/Node/React-Native (not tested yet)

Includes TypeORM, Map based and Tiered Key Value storage adapters.

Lastly there is a separate API/Interfaces within Veramo, to not expose core data types from the package.

Tests added and code "fully" implemented.

Happy to get some feedback

### CODE REVIEW

**Review by simonas-notcat — CHANGES_REQUESTED** [review]
Since this plugin is introducing top level methods, it has to have a `plugin.schema.json` file which is going to be used when generating OpenAPI schema for remote method execution. This means that the interface and arguments of these methods have to be serialisable.

You have to add this section to the `package.json` file:
```json
  "veramo": {
    "pluginInterfaces": {
      "IKeyValueStore": "./src/key-value-types.ts"
    }
  },
```

and add this line to the `scripts` section:
```json
"generate-plugin-schema": "node ../cli/bin/veramo.js dev generate-plugin-schema"
```

When you do this, you will see that this script is complaining about the types. I think the issue is that you are using `<ValueType>` in interface definitions. I don't think this is supported by the tools we are using for schema generation.

**Comment by nklomp** [comment]
Hi @simonas-notcat Thanks for the review and sorry for not getting back sooner. It has been a bit hectic over here.

I am not sure about your plugin/schema comment. This is not a Veramo plugin to be clear. It is more like the key manager, or even better the data-store. Both which are not exposing a plugin schema.

It is not a plugin, because you might want to use one or more KV stores in your own plugins. Then you want to have typed values for your keys per KV store. You are not going to call a KV Veramo instance remotely, you would use a remote backing store, like a TypeORM DB or Mongo for instance.

**Comment by simonas-notcat** [comment]
> This is not a Veramo plugin to be clear.

Sorry I did not realise that. Now it makes sense. Looks good to me

**Review by simonas-notcat — APPROVED** [review]

**Comment by nklomp** [comment]
@mirceanis Do you also want to do a review or chime in on the PR, given you had some opinions/suggestions in Discord as well?

**Review by mirceanis — COMMENTED** [review]
Looks great! There are just a few nits.

File: packages/kv-store/src/key-value-store.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    if (result === null || result === undefined) {
+      throw Error(`Result cannot be undefined or null at this this point`)
+    } else if (typeof result !== 'object') {
+      return { value: result, expires: undefined }
+    } else if (!('value' in result)) {
+      return { value: result, expires: undefined }
+    }
+    if (!('expires' in result)) {
+      result.expires = undefined
+    }
```
[the remark this answers is not in the document]
Comment (nklomp) [reply packages/kv-store/src/key-value-store.ts:118]: If I am not mistaken the original package always has the property. But indeed we don't necessarily have to do that, as we are using our own types anyway,

File: packages/kv-store/src/key-value-store.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  IKeyValueStoreOptions,
+  IValueData,
+  ValueStoreType,
+} from './key-value-types.js'
+import { Keyv } from './keyv/keyv.js'
+import { KeyvDeserializedData, KeyvOptions, KeyvStoredData } from './keyv/keyv-types.js'
+
+/**
+ * Agent plugin that implements {@link @veramo/core-types#IKeyValueStore} interface
+ * @public
```
[the remark this answers is not in the document]
Comment (nklomp) [reply packages/kv-store/src/key-value-store.ts:13]: Leftover from when I started I guess. Will remove it

[the remark this answers is not in the document]
Comment (nklomp) [reply packages/kv-store/src/key-value-store.ts]: Removed in latest commit

File: packages/cli/src/dev.ts [hunk 3]
```diff
@@ -31,7 +31,7 @@ function createSchema(generator: TJS.SchemaGenerator, symbol: string) {
     return { components: { schemas: {} } }
   }
 
-  let fixedSymbol = symbol.replace('Array<', '').replace('>', '')
+  let fixedSymbol = symbol.replace(/Array\<(.*)\>/gm, '$1')
```
Comment (nklomp) [inline packages/cli/src/dev.ts:34]: @mirceanis Just wanted to highlight this change outside of the kv-store package. When testing a bit with the schema-generator I noticed that it mangled anything with carets in there, whilst it was only intended for Arrays.

**Comment by nklomp** [comment]
Okay comments are addressed.

2 things:
- I took the liberty to add myself to the list of Authors
- See the remark about the change in the CLI package

**Comment by mirceanis** [comment]
> * I took the liberty to add myself to the list of Authors

That's exactly how it should be :)
We value your contributions greatly!

**Pull request merged**